engine/src/main/java/io/camunda/zeebe/engine/state/migration/DbMigratorImpl.java

```diff
@@ -0,0 +1,96 @@
+/*
+ * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH under
+ * one or more contributor license agreements. See the NOTICE file distributed
+ * with this work for additional information regarding copyright ownership.
+ * Licensed under the Zeebe Community License 1.1. You may not use this file
+ * except in compliance with the Zeebe Community License 1.1.
+ */
+package io.camunda.zeebe.engine.state.migration;
+
+import io.camunda.zeebe.engine.state.mutable.MutableZeebeState;
+import java.util.Collections;
+import java.util.List;
+import java.util.function.Supplier;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class DbMigratorImpl implements DbMigrator {
+
+  private static final Logger LOGGER =
+      LoggerFactory.getLogger(DbMigratorImpl.class.getPackageName());
+
+  // add new migration tasks here, migrations are executed in the order they appear in the list
+  private static final List<MigrationTask> MIGRATION_TASKS = Collections.emptyList();
+
+  private final MutableZeebeState zeebeState;
+  private final Supplier<List<MigrationTask>> migrationSupplier;
+  private boolean abortRequested = false;
+
+  public DbMigratorImpl(final MutableZeebeState zeebeState) {
+    this(zeebeState, () -> MIGRATION_TASKS);
+  }
+
+  DbMigratorImpl(
+      final MutableZeebeState zeebeState, final Supplier<List<MigrationTask>> migrationSupplier) {
+
+    this.zeebeState = zeebeState;
+    this.migrationSupplier = migrationSupplier;
+  }
+
+  @Override
+  public void runMigrations() {
+    LOGGER.info("Starting migrations ...");
+    final var migrationTasks = migrationSupplier.get();
+
+    for (int index = 1; index <= migrationTasks.size(); index++) {
+      // one based index looks nicer in logs
+      handleMigrationTask(migrationTasks.get(index - 1), index, migrationTasks.size());
+    }
+    if (!abortRequested) {
+      LOGGER.info("Finished running migrations");
+    }
+  }
+
+  @Override
+  public void abort() {
+    LOGGER.info("Received abort signal");
+    abortRequested = true;
+  }
+
+  private void handleMigrationTask(
+      final MigrationTask migrationTask, final int index, final int total) {
+    if (abortRequested) {
+      return;
+    }
+
+    if (migrationTask.needsToRun(zeebeState)) {
+      runMigration(migrationTask, index, total);
+    } else {
+      logMigrationSkipped(migrationTask, index, total);
+    }
+  }
+
+  private void logMigrationSkipped(
+      final MigrationTask migrationTask, final int index, final int total) {
+    LOGGER.info(
```

Comment:
So on every failover once I've migrated, this would be logged?

Reply:
On every failover, on every start, on every transition to leader (which triggers recovery)